### ISSUE TITLE
feature request: file.recurse could copy file permissions if not explicitly set with file_mode

### ISSUE THREAD

**Issue by Mrten** [saltstack/salt#4423]
After a rather unsuccessfull poll on the mailinglist, I'll leave this here:

If I have

```
/usr/share/munin/plugins:
  file.recurse:
    - source: salt://all-munin-node/plugins
    - clean: false  # just in case, anders zijn alle plugins weg
    - require:
      - pkg: munin-node-pkgs
```

and the plugins/ directory looks like this:

```
root@salt-master:/home/salt/conf/all-munin-node/plugins# ls -aslF
total 16
0 drwxr-xr-x 2 root root   61 Apr  6 10:49 ./
0 drwxr-xr-x 4 root root   92 Apr  6 10:52 ../
4 -rwxr-xr-x 1 root root 2966 Apr  6 10:48 kvm_cpu*
4 -rwxr-xr-x 1 root root 3307 Apr  6 10:48 kvm_io*
4 -rwxr-xr-x 1 root root 3201 Apr  6 10:48 kvm_mem*
4 -rwxr-xr-x 1 root root 4004 Apr  6 10:49 kvm_net*
```

wouldn't it be reasonable to expect that the files are created 755 on
the minion?

Because they aren't:

```
root@0010:/usr/share/munin/plugins# ls -aslF | grep kvm
 4 -rw-r--r-- 1 root  root   2966 Apr  6 10:52 kvm_cpu
 4 -rw-r--r-- 1 root  root   3307 Apr  6 10:52 kvm_io
 4 -rw-r--r-- 1 root  root   3201 Apr  6 10:52 kvm_mem
 4 -rw-r--r-- 1 root  root   4004 Apr  6 10:52 kvm_net
```

I know I can set `file_mode`, but that sets the mode of _all_ files to
something, while it could be beneficial to have some files 644, other
400 and some 755.

Perhaps a new setting is in order: "copy_permissions"

M.

**Comment by Cottser** [comment]
Just searching around, might this be a duplicate of https://github.com/saltstack/salt/issues/15784?

**Comment by Mrten** [comment]
nah, this one should be able to mix rwx, r-x and --x for different files in the same directory.

**Comment by Cottser** [comment]
Carry on then :)

**Comment by uvsmtid** [comment]
I agree with @Mrten. Let's keep this issue open until Salt is able to transfer detailed permission information about each file/directory from the source. This is important because source can potentially list 10, 100, 1000, ... files and directories. It's impractical to code (!) these details in Salt state file if you want to replicate Read/Write/Executable fields per User/Group/Other for all files _exactly_.

I believe that majority of cases require only Executable (or not) for All (or none) - single flag just to keep scripts executable. But I also believe that, based on @thatch45 comment above, as soon as additional metadata about file permissions for file transfer is added, making it complete right away should not cost much.

If someone seriously thinks to suggest a workaround by `tar`-ing these files/directories in Salt master, pushing archive to Salt minion, and extracting it there, he didn't try implementing this cycle himself.

**Comment by jfindlay** [comment]
If this feature is going to be implemented, I suggest also including extended attributes to the extent they make sense and are supported.

**Comment by babilen** [comment]
Does https://github.com/saltstack/salt/pull/34807 address this?

**Comment by Mrten** [comment]
probably...

**Comment by UtahDave** [comment]
Closed by #34807
